### PR TITLE
INSP:added inspection for dyn keyword for trait objects

### PR DESCRIPTION
**Pull request by rrevenantt**
related  to #2622

### CODE REVIEW

**Comment by vlad20012** [comment]
Wow, nice! I recently thought about creating a "help wanted" issue about it :)

You're missed a case of `RsTraitType` without `dyn` keyword, e.g. `&(A+B)` instead of `&(dyn A+B)`

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt [hunk 1]
```diff
@@ -427,6 +427,10 @@ class RsPsiFactory(
         createExpression(text) as? E
             ?: error("Failed to create ${E::class.simpleName} from `$text`")
 
+    fun createTraitType(pathText: String) =
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt]:
```suggestion
    fun createTraitType(pathText: String): RsTraitType =
```

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt [hunk 2]
```diff
@@ -427,6 +427,10 @@ class RsPsiFactory(
...
+        createFromText<RsTraitType>("fn a(){let a:&dyn $pathText;}")
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt]:
```suggestion
        createFromText<RsTraitType>("type T = &dyn $pathText;")
```

File: src/main/kotlin/org/rust/ide/inspections/RsTraitObjectInspection.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.rust.lang.core.psi.ext.isEdition2018
+import org.rust.lang.core.resolve.ref.deepResolve
+
+class RsTraitObjectInspection : RsLocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object : RsVisitor() {
+        override fun visitTypeReference(typeReference: RsTypeReference) {
+            if (!typeReference.isEdition2018) return
+
+            if (typeReference.baseType?.path?.reference?.deepResolve() is RsTraitItem
+                || typeReference.traitType?.firstChild?.elementType == RsElementTypes.POLYBOUND
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/RsTraitObjectInspection.kt]: Looks too hacky. Just add this
```
val RsTraitType.dyn: PsiElement? get() = node.findChildByType(RsElementTypes.DYN)?.psi
```
to `RsTraitType.kt`. Then use it here:

```suggestion
                || typeReference.traitType?.dyn == null
```

Also could you please ping your reviewer after resolving comments, b/c I don't receive any notifications on force-pushes to a PR, and it's hard to track them manually.

**Comment by rrevenantt** [comment]
@vlad20012 `RsCompletionAutoPopupTest > test path auto popup FAILED` on clion-192 in travis. Weird. I have no idea how my changes can affect this test. And it does not look like some random test fail because it failed both after empty forse-push and after rebase.

**Comment by Undin** [comment]
@rrevenantt your changes don't affect it. For some reason, this test is blinking. We are investigating why. I've restarted failed build

**Comment by vlad20012** [comment]
:+1: 
bors r+

**Pull request merged**